File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res, sol = [], []
        n = len(nums)

        def backtrack(i, summ):
            if summ > target:
                return 
            
            if summ == target:
                res.append(sol[:])
                return 

            for j in range(i, n):
                sol.append(nums[j])
                backtrack(j, summ + nums[j])
                sol.pop()

        backtrack(0, 0)
        return res

File: Data Structures & Algorithms/permutations/submission-1.go
func permute(nums []int) [][]int {
	sol := []int {} 
	res := [][]int {}
	used := make([]bool, len(nums))

	var backtrack func(i int) 

	backtrack = func(i int) {
		if i == len(nums) {
			res = append(res, append([]int{}, sol...))
			return 
		}

		for j := 0; j < len(nums); j++ {
            if used[j] {
				continue
			}
			
			used[j] = true
			sol = append(sol, nums[j])
			backtrack(i + 1)
			sol = sol[:len(sol) - 1]
			used[j] = false
		}
	}

	backtrack(0)
	return res
}
